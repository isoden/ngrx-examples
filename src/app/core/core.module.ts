import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';

import { AppComponent } from './containers/app/app.component';
import { TopPageComponent } from './containers/top-page/top-page.component';

@NgModule({
  imports: [
    CommonModule,
    RouterModule.forChild([
      {
        path     : '',
        component: TopPageComponent,
      },
    ]),
  ],
  declarations: [
    AppComponent,
    TopPageComponent,
  ],
})
export class CoreModule { }
