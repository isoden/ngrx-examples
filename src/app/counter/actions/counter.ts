import { Action } from '@ngrx/store';

export enum CounterActionTypes {
  Increment = '[Counter] Increment',
  Decrement = '[Counter] Decrement',
}

export class Increment implements Action {
  readonly type = CounterActionTypes.Increment;
}

export class Decrement implements Action {
  readonly type = CounterActionTypes.Decrement;
}

export type CounterActions =
  | Increment
  | Decrement;
