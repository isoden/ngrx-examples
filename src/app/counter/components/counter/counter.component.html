<pre><code>counter =&gt; {{ counter }}</code></pre>

<button (click)="increment.emit()">increment</button>
<button (click)="decrement.emit()">decrement</button>
<button (click)="incrementAsync()">increment async</button>
<button (click)="incrementIfOdd()">increment if odd</button>
