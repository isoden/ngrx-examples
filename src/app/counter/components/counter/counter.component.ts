import {
  Component,
  OnInit,
  ChangeDetectionStrategy,
  Input,
  Output,
  EventEmitter,
} from '@angular/core';

@Component({
  selector: 'app-counter',
  templateUrl: './counter.component.html',
  styleUrls: ['./counter.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class CounterComponent implements OnInit {
  /**
   * counter value
   */
  @Input()
  counter: number;

  /**
   * EventEmitter for increment event
   */
  @Output()
  increment = new EventEmitter<void>();

  /**
   * EventEmitter for decrement event
   */
  @Output()
  decrement = new EventEmitter<void>();

  constructor() { }

  ngOnInit() {
  }

  /**
   * emit increment event after 1000ms
   */
  incrementAsync() {
    setTimeout(() => this.increment.emit(), 1000);
  }

  /**
   * if a counter value is odd, emit increment event
   */
  incrementIfOdd() {
    if (this.counter % 2 !== 0) {
      this.increment.emit();
    }
  }
}
