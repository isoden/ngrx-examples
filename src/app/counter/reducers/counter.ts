import { CounterActions, CounterActionTypes } from '../actions/counter';

export type CounterState = number;
const initialState: CounterState = 0;

export function reducer(state = initialState, action: CounterActions): CounterState {
  switch (action.type) {
    case CounterActionTypes.Increment: return state + 1;
    case CounterActionTypes.Decrement: return state - 1;

    default: return state;
  }
}

export const getCounter = (state: CounterState) => state;
