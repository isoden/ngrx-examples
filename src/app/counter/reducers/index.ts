import { createSelector, createFeatureSelector } from '@ngrx/store';
import * as fromCounter from './counter';

export const reducers = fromCounter.reducer;

export interface State {
  counter: fromCounter.CounterState;
}

export const selectCounter = createFeatureSelector<fromCounter.CounterState>('counter');
export const getCounter = createSelector(selectCounter, fromCounter.getCounter);
