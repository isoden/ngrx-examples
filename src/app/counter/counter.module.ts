import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { StoreModule } from '@ngrx/store';
import { CounterPageComponent } from './containers/counter-page/counter-page.component';
import { CounterComponent } from './components/counter/counter.component';
import { reducers } from './reducers';

@NgModule({
  imports: [
    CommonModule,
    StoreModule.forFeature('counter', reducers),
    RouterModule.forChild([
      {
        path     : '',
        component: CounterPageComponent,
      }
    ])
  ],
  declarations: [CounterPageComponent, CounterComponent]
})
export class CounterModule { }
