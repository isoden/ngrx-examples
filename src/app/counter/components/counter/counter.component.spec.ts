import { async, ComponentFixture, TestBed, fakeAsync, tick } from '@angular/core/testing';
import { By } from '@angular/platform-browser';

import { CounterComponent } from './counter.component';
import { DebugElement } from '@angular/core/src/debug/debug_node';
import { ChangeDetectionStrategy } from '@angular/core';

describe('CounterComponent', () => {
  let component: CounterComponent;
  let fixture: ComponentFixture<CounterComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [ CounterComponent ]
    })
    .overrideComponent(CounterComponent, {
      set: { changeDetection: ChangeDetectionStrategy.Default },
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CounterComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should show current counter', async(() => {
    component.counter = 13;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  }));

  it('should emit "increment"', () => {
    const button = fixture.debugElement.query(By.css('button:nth-of-type(1)'));

    jest.spyOn(component.increment, 'emit');
    button.triggerEventHandler('click', null);

    expect(component.increment.emit).toHaveBeenCalled();
  });

  it('should emit "decrement"', () => {
    const button = fixture.debugElement.query(By.css('button:nth-of-type(2)'));

    jest.spyOn(component.decrement, 'emit');
    button.triggerEventHandler('click', null);

    expect(component.decrement.emit).toHaveBeenCalled();
  });

  it('should emit "increment" async', fakeAsync(() => {
    const button = fixture.debugElement.query(By.css('button:nth-of-type(3)'));

    jest.spyOn(component.increment, 'emit');
    button.triggerEventHandler('click', null);

    expect(component.increment.emit).not.toHaveBeenCalled();

    tick(1000);

    expect(component.increment.emit).toHaveBeenCalled();
  }));

  describe('incrementIfOdd', () => {
    let button: DebugElement;

    beforeEach(() => {
      button = fixture.debugElement.query(By.css('button:nth-of-type(4)'));
      jest.spyOn(component.increment, 'emit');
    });

    it('should not emit "increment" if counter is even', () => {
      component.counter = 0;

      button.triggerEventHandler('click', null);

      expect(component.increment.emit).not.toHaveBeenCalled();
    });

    it('should emit "increment" if counter is odd', () => {
      component.counter = 1;

      button.triggerEventHandler('click', null);

      expect(component.increment.emit).toHaveBeenCalled();
    });
  });

});
