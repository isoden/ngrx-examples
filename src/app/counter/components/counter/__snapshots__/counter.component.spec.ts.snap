// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterComponent should show current counter 1`] = `
<app-counter
  counter={[Function Number]}
  decrement={[Function EventEmitter]}
  increment={[Function EventEmitter]}
>
  <pre>
    <code>
      counter =&gt; 13
    </code>
  </pre><button>
    increment
  </button><button>
    decrement
  </button><button>
    increment async
  </button><button>
    increment if odd
  </button>
</app-counter>
`;
